primitives/messages/src/lib.rs

```diff
@@ -483,20 +486,35 @@ pub enum ReceivalResult<DispatchLevelResult> {
 	TooManyUnconfirmedMessages,
 }
 
+/// Type of reward that needs to be paid at the source chain for delivering messages to the
+/// target chain. It does not necessary mapped 1:1 onto source chain tokens (or whatever is
+/// used to pay rewards there). The `DeliveryConfirmationPayments` implementation decides how
+/// to map it and what actual reward the relayer would receive.
+///
+/// Why our code is not generic over this type? That's mostly because we are using single
+/// `pallet-bridge-relayers` instance to register rewards for all bridges. Those bridges
+/// may be connected to chains, using different `Balance` types. Why not to use encoded version?
+/// Because we need to order relayers by the reward they get for delivering a single message.
+pub type RewardAtSource = u64;
+
 /// Delivered messages with their dispatch result.
 #[derive(Clone, Default, Encode, Decode, RuntimeDebug, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
 pub struct DeliveredMessages {
 	/// Nonce of the first message that has been delivered (inclusive).
 	pub begin: MessageNonce,
 	/// Nonce of the last message that has been delivered (inclusive).
 	pub end: MessageNonce,
+	/// Reward that needs to be paid at the source chain (during confirmation transaction)
+	/// for **every delivered message** in the `begin..=end` range. If reward has been paid
+	/// at the target chain or if no rewards assumed, it may be zero.
+	pub reward_per_message: RewardAtSource,
```

Comment:
> Just trying to understand how will this be computed ? ... Or will it be set by the target chain somehow ?

Yes, it is specified by the relayer. So the current idea (#2605) assumes that we will have a bounded circular queue of relayers, which are getting priority boosts for message delivery transactions in their slots. So if queue is `[a, b, c]`, then relayer `a` may get priority for transactions that are **verified** at blocks `1..100`, relayer `b` - at blocks `100..200`, relayer `c` - at blocks `200..300`, then relayer `a` again at blocks `300..400` and so on.

Then there's a question - how to fill this queue? If maximal number of relayers in the queue is `N` and there are lte than `N` relayers, then we allow all relayers to serve at lane.  But if there are more than `N` relayers that are willing to get additional priority boost, then we need to choose. Luckily, one of our goals in this protocol is to (try to) make messages cheaper. Message fee that is paid at the sending chain is the transport fee + dispatch fee. Part of this transport fee is the rewards that we are paying to relayers (that's how we incentivize relayers). So we want to incentivize relayers, but at the same time we want to make reward smaller, which shall eventually make message fee (paid by the sender) smaller too.

So we need some kind of market here, where relayers are offering their services, specifying reward that they want to receive. If we already have `N` relayers, but then next relayer comes, which is ready to get smaller reward for delivering messages, we will allow him to do that. So he'll take his place in the queue, thus some "greedier" relayer will be pushed out of the queue.

> Is this an amount requested by the relayer for delivering the message ?

In practice - yes, we will probably start with that, because we are bridging similar chains. But at the same time, we can't force all chains to use this 1:1 mapping. Even for P<>K it may be impossible, because they use `u128` for native tokens balance and here we use `u64`. So e,g. paying large rewards is impossible (I hope we are not going to pay that large rewards, though). For other chains it may be mapped e.g. to the percent of maximal possible reward that the relayer wants. I.e. relayer who wants `10%` of max reward would win vs relayer who wants `60%` (perhaps we will also should start with that?).

That said, the actual reward is always computed by the `calc_relayers_rewards_at_source` at the source chain, where the this value is given some real-world meaning. Until that, it is only used to sort relayers at the target chain.

Reply:
Thank you for the details. Makes sense. So if this is the amount requested by the relayer, could we rename it to something like `relayer_fee_per_message` ? And also is `DeliveryPayments::delivery_reward_per_message()` needed: https://github.com/paritytech/parity-bridges-common/blob/6ed9f0fababd666c93880ee4b374d28c033bed9c/modules/messages/src/lib.rs#L273 ?